### ISSUE TITLE
Issue 5: Implement strict leakage-free temporal evaluation for SCADANet and BATADAL

### ISSUE THREAD

**Issue by Mati86** [AI-Security-Internships-2026/12-graph-based-cyber-physical-risk#26]
# Issue 5: Implement strict leakage-free temporal evaluation for SCADANet and BATADAL

## Milestone: M3 — Leakage-Free Temporal and Shortcut Robustness

---

## Problem Statement

Replace the current approximate chronological evaluation with a **strict leakage-free temporal protocol**.

The central paper argues that evaluation methodology can create misleadingly strong results. Therefore the revised temporal protocol must itself avoid future-information leakage.

---

## Research Questions Supported

**RQ1:** How much do evaluation protocols inflate reported performance?  
**RQ3:** Do the shortcut/generalization findings remain under realistic chronological evaluation?

---

#### Problem to fix

The previous temporal experiment uses chronological train/test ordering, but later test-period graph edges can be available to predictions for earlier test examples.

This is not strict online evaluation.

#### Required Protocol

Use time-ordered windows.

For each test window `t`:

```text
training-period graph
       +
topology observed before test window t
       ↓
predict flows in window t
       ↓
after evaluation, update graph with window t
       ↓
move to window t+1
```

At prediction time, **no edge, node event, or feature derived from future windows may be available**.

##### Important design choice

Document whether the model itself is:

- frozen after training, or
- incrementally updated.

For the main paper experiment, prefer:

> **Frozen model + evolving observed topology**

because this isolates temporal generalization from continual-learning effects.

Continual learning can be future work.

##### SCADANet temporal variants

Run:

###### T1 — Existing chronological protocol

Keep only as historical reference.

###### T2 — Strict cumulative-topology protocol

No future test edges.

###### T3 — Optional fixed-training-topology protocol

Use only topology known at the end of training for every future test flow.

This gives a useful comparison:

- no topology update,
- causal topology update.

##### Required SCADANet diagnostics

For each test window record:

- number of flows,
- attack rate,
- attack-type counts,
- new nodes,
- new IP pairs,
- cumulative graph nodes/edges,
- F1,
- AUPRC,
- recall,
- FPR,
- per-attack recall.

Especially track:

- `vuln_scan`,
- `modbus_fdi`,
- `modbus_abuse`,
- `insider_threat`,
- `apt_exfil`.

#### Problem to fix

Current BATADAL windows are:

```text
window size = 24 hours
stride = 6 hours
```

Adjacent windows overlap by 18 hours.

A train/test boundary can therefore contain raw observations shared by both sides.

#### Required Protocol

Implement a **purged / embargoed chronological split**.

Example:

```text
TRAIN windows
───────────────
PURGE / EMBARGO
───────────────
TEST windows
```

No raw timestamp/row used by a training window may appear in any test window.

The simplest correct method is acceptable.

Document:

- number of windows removed by purge,
- resulting train/test counts,
- number of anomalous windows in each split.

###### T4 — Existing overlapping temporal split

Historical reference only.

###### T5 — Purged 70/30 chronological split

Primary result.

###### T6 — Split sensitivity

Repeat with at least:

- 60/40
- 70/30
- 80/20

while applying the purge rule consistently.

Threshold selection must use training/validation data only.

### 3. Result Artifacts

Create:

```text
experiments/results/q1/temporal/
```

Suggested files:

```text
scadanet_temporal_window_metrics.csv
scadanet_temporal_summary.csv
batadal_purged_temporal_metrics.csv
batadal_boundary_audit.json
```

#### Table T1 — Old vs strict temporal evaluation

| Dataset | Protocol | Precision | Recall | F1 | AUPRC | FPR | Leakage condition |
|---|---|---:|---:|---:|---:|---:|---|
| SCADANet | Previous temporal | | | | | | future test topology possible |
| SCADANet | Strict causal topology | | | | | | none identified |
| BATADAL | Previous temporal | | | | | | overlapping raw rows |
| BATADAL | Purged temporal | | | | | | none identified |

#### Table T2 — SCADANet temporal performance per attack

| Attack type | Train count | Test count | Static/Track-B recall | Strict temporal recall | Δ recall |
|---|---:|---:|---:|---:|---:|

#### Table T3 — BATADAL split sensitivity after purging

| Split | Train windows | Purged windows | Test windows | Precision | Recall | F1 | AUPRC |
|---|---:|---:|---:|---:|---:|---:|---:|

#### Figure T1 — Performance over time

For SCADANet:

- x-axis = chronological test window,
- plot F1,
- recall,
- FPR or AUPRC.

Prefer separate files if one figure becomes crowded.

#### Figure T2 — Graph growth over time

Plot:

- cumulative nodes,
- cumulative unique IP pairs/edges,
- indicate train/test boundary.

This shows how much topology is unseen at training time.

#### Figure T3 — Attack prevalence over time

Stacked or multi-line plot of major attack types by window.

Highlight `vuln_scan`.

#### Figure T4 — Static vs strict temporal per-attack recall

Heatmap or paired dot plot.

This should be a **main paper figure**.

#### Figure T5 — BATADAL purge illustration

A simple conceptual figure showing overlapping 24h windows and which boundary windows are removed.

This is useful for explaining the leakage correction.

---

## Expected Research Result

The important question is:

> Does the temporal generalization gap remain after removing temporal leakage?

Possible outcomes:

### Outcome A — gap remains

This strongly supports the paper:

> Static evaluation substantially overestimates future-time performance even under strict causal evaluation.

### Outcome B — performance changes materially after leakage removal

Also important:

> Previously reported temporal metrics were themselves affected by information overlap, demonstrating why temporal protocol design matters.

### Outcome C — specific attack classes recover/fail

This can reveal that the temporal issue is attack-specific rather than universal.

---

## Acceptance Criteria

- [ ] Test samples are processed chronologically.
- [ ] No future SCADANet test edges are visible to earlier predictions.
- [ ] The graph state used for each test window is documented/saved.
- [ ] BATADAL train/test raw observations do not overlap.
- [ ] Purged window count is reported.
- [ ] Threshold/hyperparameter tuning does not use test labels.
- [ ] Old and strict protocols are both reported for comparison.
- [ ] Per-window SCADANet metrics are saved.
- [ ] Per-attack strict temporal recall is reported.
- [ ] Table T1–T3 and Figure T1–T5 can be generated directly from committed artifacts.

---

## Do Not Do In This Issue

Do not:

- update the model using future test labels,
- let later test topology leak into earlier predictions,
- select threshold using final test performance,
- silently discard difficult time periods.

The strict protocol is more important than obtaining high performance.
